src/firmware/main/app/control.c

```diff
@@ -0,0 +1,182 @@
+#include "app/control.h"
+
+#include "physics/physics.h"
+#include "shared/util.h"
+
+/**
+ * Computes the scaling constant to bring the wheel forces to their maximum
+ *
+ * Note that this could scale the forces *down* if they exceed the physical capabilities
+ * of the robot (ex. wheel slip).
+ *
+ * @param wheels Wheels to compute the maximum torque scaling from
+ * @param wheel_forces Forces to apply to each wheel
+ * @param battery_voltage The current battery voltage
+ *
+ * @return The amount by which to scale the force on each wheel to get the maximum
+ *         torque possible while maintaining the same torque ratio between wheels
+ */
+float app_control_getMaximalTorqueScaling(const Wheel_t* wheels[4],
+                                          const float wheel_forces[4],
+                                          float battery_voltage)
+{
+    float max_effective_motor_voltage = -INFINITY;
+    float slip_ratio_min              = INFINITY;
+
+    for (long i = 0; i < 4; i++)
+    {
+        Wheel_t* wheel                   = wheels[i];
+        const WheelConstants_t constants = app_wheel_getWheelConstants(wheel);
+        float force                      = wheel_forces[i];
+        float motor_torque =
+            force * constants.wheel_radius * constants.wheel_rotations_per_motor_rotation;
+        float curr_motor_rpm = app_wheel_getMotorSpeedRPM(wheel);
+
+        float applied_voltage = motor_torque * constants.motor_current_per_unit_torque *
+                                constants.motor_phase_resistance;
+        float back_emf          = curr_motor_rpm * constants.motor_back_emf_per_rpm;
+        float effective_voltage = fabsf(applied_voltage + back_emf);
+
+        float slip_ratio =
+            constants.motor_max_delta_voltage_before_wheel_slip / fabsf(applied_voltage);
```

Comment:
Why are we not using a delta voltage here to calculate the chance of slipping like the name of the constant implies?

Reply:
I'm confused, we are?

Reply:
Ah, absolutely correct. More fun copyover from the old firmware. I'll change to `motor_max_voltage_before_wheel_slip` and update the constant it comes from as well...........